integrations/gcp/gcp_core/helpers/ratelimiter/base.py

```diff
@@ -0,0 +1,178 @@
+            )
+            return self._default_quota
+
+    @abstractmethod
+    def quota_name(self, container_id: str) -> str:
+        """Generate the fully qualified name for the quota resource."""
+        pass
+
+
+class GCPResourceRateLimiter(GCPResourceQuota):
+    """
+    GCPResourceRateLimiter manages rate limits based on GCP resource quotas.
+    It inherits from GCPResourceQuota and leverages GCP's quota information to dynamically determine and apply rate limits using the AsyncLimiter.
+    This class allows efficient control over the rate of API requests per container (e.g., projects, folders) based on the actual quota allocated by GCP.
+    """
+
+    time_period: float = _DEFAULT_RATE_LIMIT_TIME_PERIOD
+
+    @cache_coroutine_result()
+    async def register(self, container_id: str, *arg: Optional[Any]) -> AsyncLimiter:
+        quota = await self._get_quota(container_id, *arg)
+
+        effective_quota_limit: int = int(max(round(quota * _PERCENTAGE_OF_QUOTA, 1), 1))
+        logger.info(
+            f"The Integration will utilize {_PERCENTAGE_OF_QUOTA * 100}% of the quota, which equates to {effective_quota_limit} for rate limiting."
+        )
+
+        limiter = AsyncLimiter(max_rate=quota, time_period=self.time_period)
+        limiter.container_id = container_id
+        return limiter
+
+    @final
+    def quota_name(self, container_id: str) -> str:
+        return f"{self.container_type.value}/{container_id}/locations/global/services/{self.service}/quotaInfos/{self.quota_id}"
+
+    async def limiter(self, container_id: str) -> AsyncLimiter:
+        "fetches the rate limiter for the given container"
+        name = self.quota_name(container_id)
+        return await self.register(container_id, name)
+
+
+class ResourceBoundedSemaphore(GCPResourceRateLimiter):
+    """
+    scope: The container (project, organization or folder) within which the service account was created.
+    Implement access to the scope via container_id class method.
+    The class inherits from GCPResourceRateLimiter and provides a semaphore that can be used to control the number of containers queried concurrently for a given service and quota based on the base container's quota.
+    """
+
+    _semaphore: Optional[asyncio.BoundedSemaphore] = None
+    default_maximum_limit: int = _MAXIMUM_CONCURRENT_REQUESTS
+
+    async def semaphore(self, container_id: str) -> asyncio.BoundedSemaphore:
+        if self._semaphore is None:
+            _maximum_concurrent_requests = await self._maximum_concurrent_requests(
+                container_id
+            )
+            self._semaphore = asyncio.BoundedSemaphore(_maximum_concurrent_requests)
+        return self._semaphore
+
+    async def _maximum_concurrent_requests(self, container_id: str) -> int:
+        name = self.quota_name(container_id)
+        quota = await self.register(container_id, name)
+        limit = max(
```

Comment:
Could you please explain how did you come to this calculation here?

Reply:
This gives an estimate of how many projects can be allowed to run within the `quota's limits` .

The higher the number of concurrent requests, the lesser the number of resources each project gets to query at a time before the quota gets exhausted. If the number of projects attached to the service account is relatively large, it means the integration will take a longer before ingesting resources to port since each of those projects would have to finish paginating through their resources (kind) first before an ingestion happens. However, by controlling this value, we directly control our ingestion `frequency` by simply obeying the quota of the limiter.




   - The `max` function is used to determine the larger value between the calculated limit and the `self.default_maximum_limit`.
   - This ensures that the system either uses the calculated limit (if it's greater) or defaults to the `self.default_maximum_limit` (if the calculated limit is smaller).

### Example:
Let's say the quota's maximum rate (`quota.max_rate`) is 320, and the `self.default_maximum_limit` is set to 10.

   - `limit = floor(320 / 10) = 32`
   - `semaphore = max(32, 10) = 32`

### Explanation:
In this case, the calculated limit is 32, which is greater than the default maximum limit of 10. Hence, the integration will permit a batch of `32 projects to battle for 320 requests` at a time. This increases the chance of us completing a sync for some projects within the batch. 
Note that for a sync to happen, all pages of a project's resource must be iterated through, it could be any number of pages though, but counts of all the requests needed to complete the iterations are deducted from the 320 for all projects within the batch.

Reply:
@matan84  makes sense ?

Reply:
Yup, makes sense